bde/ml/models.py

```diff
@@ -827,17 +897,137 @@ def _prior_fitting(
         return params, history
 
-    def mcmc_loop(self):
-        r"""Perform MCMC sampling for BDE estimator."""
-        ...
+    def mcmc_sampling(
+        self,
+        model_states: train_state.TrainState,
+        rng_key: PRNGKeyArray,
+        train: datasets.BasicDataset,
+    ) -> List[Dict]:
+        r"""Perform MCMC-burn-in and sampling."""
+        sample_keys = jax.random.split(rng_key, self.n_chains + 1)
+        split_key, sample_keys = sample_keys[0], sample_keys[1:]
+
+        @jax.jit
+        def logdensity_for_batch(
+            params,
+            carry: float,
+            batch: Tuple[ArrayLike, ArrayLike],
+        ) -> Tuple[float, None]:
+            x, y = batch
+            y_pred = self.model_.apply(params, x)
+            res_loss = self.loss(y, y_pred)
+            res_loss = jnp.sum(res_loss)
+            res_prior = self.log_prior(params)
+            return carry + res_loss + res_prior, None
+
+        @jax.jit
+        def logdensity(params):
+            res, _ = jax.lax.scan(
+                f=lambda carry, batch: logdensity_for_batch(
+                    params=params,
+                    carry=carry,
+                    batch=batch,
+                ),
+                init=0.0,
+                xs=train.get_scannable(),
+            )
+            return res
+
+        warmup = blackjax.window_adaptation(
+            blackjax.nuts,
+            logdensity,
+            # progress_bar=True,
+        )
+
+        @partial(jax.jit, static_argnums=3)
+        def burn_in_loop(
+            rng: PRNGKeyArray,
+            params: ArrayLike,
+            data: datasets.BasicDataset,
+            n_burns: int,
+        ):
+            # NOTE: We require the `data` variable despite not using it
+            #  to ake the jit tracing work properly.
+            return warmup.run(rng, params, n_burns)
+
+        (init_state_sampling, nuts_params), warmup_info = jax.pmap(
+            fun=burn_in_loop,
+            in_axes=(0, 0, None, None),
+            static_broadcasted_argnums=(3,),
+        )(
+            sample_keys,
+            model_states.params,
+            train,
+            self.warmup,  # num_steps
+        )
+
+        @partial(jax.jit, static_argnames=["num_samples"])
+        def inference_loop(
+            rng_key,
+            initial_state,
+            nuts_params,
+            num_samples,
+        ) -> SamplingAlgorithm:
+            kernel = blackjax.nuts(logdensity, **nuts_params).step
+
+            @jax.jit
+            def one_step(state, rng_key):
+                state, _ = kernel(rng_key, state)
+                return state, state
+
+            keys = jax.random.split(rng_key, num_samples)
+            _, states = jax.lax.scan(
+                f=one_step,
+                init=initial_state,
+                xs=keys,
+            )
+
+            return states
+
+        sample_keys = jax.random.split(split_key, self.n_chains)
+        pmap_states = jax.pmap(
```

Comment:
This only works as long as n_chains < available devices right?

Reply:
This is not always the setting

Reply:
This issue should be fixed for the `fit` method in 5cf7e9e2f55edc080fc7ca20f857bdb22c807b60.
This fix raised a new efficiency issue in cases where `n_chains % n_devices > 0` which was only partially addressed, but we could document it as a bug for now and address it later based on time constraints.
I currently disabled parallelization for the `predict` methods altogether since jitting my solution for the `fit` method proved non-trivial. I plan to revisit it in a future PR.